python/cuda_parallel/cuda/parallel/experimental/__init__.py

```diff
@@ -156,50 +145,92 @@ def _host_array_to_value(array):
 class _Op:
     def __init__(self, dtype, op):
         value_type = numba.from_dtype(dtype)
-        self.ltoir, _ = cuda.compile(op, sig=value_type(
-            value_type, value_type), output='ltoir')
-        self.name = op.__name__.encode('utf-8')
+        self.ltoir, _ = cuda.compile(
+            op, sig=value_type(value_type, value_type), output="ltoir"
+        )
+        self.name = op.__name__.encode("utf-8")
 
     def handle(self):
-        return _CCCLOp(_CCCLOpKindEnum.STATELESS, self.name, ctypes.c_char_p(self.ltoir), len(self.ltoir), 1, 1, None)
-
-
-def _extract_ctypes_ltoirs(numba_cuda_compile_results):
-    view_lst = [_CCCLStringView(ltoir, len(ltoir))
-                for ltoir in numba_cuda_compile_results]
-    view_arr = (_CCCLStringView * len(view_lst))(*view_lst)
-    return ctypes.pointer(_CCCLStringViews(view_arr, len(view_arr)))
-
-
-def _facade_iter_as_cccl_iter(d_in):
-    def prefix_name(name):
-        return (d_in.prefix + "_" + name).encode('utf-8')
-    # type name ltoi ltoir_size size alignment state
-    adv = _CCCLOp(_CCCLOpKindEnum.STATELESS, prefix_name("advance"), None, 0, 1, 1, None)
-    drf = _CCCLOp(_CCCLOpKindEnum.STATELESS, prefix_name("dereference"), None, 0, 1, 1, None)
-    info = _numba_type_to_info(d_in.ntype)
-    ltoirs = _extract_ctypes_ltoirs(d_in.ltoirs)
-    # size alignment type advance dereference value_type state ltoirs
-    return _CCCLIterator(d_in.size, d_in.alignment, _CCCLIteratorKindEnum.ITERATOR, adv, drf, info, d_in.state_c_void_p, ltoirs)
-
-
-def _d_in_as_cccl_iter(d_in):
-    if hasattr(d_in, 'ntype'):
-        return _facade_iter_as_cccl_iter(d_in)
-    assert hasattr(d_in, 'dtype')
-    return _device_array_to_pointer(d_in)
+        return _CCCLOp(
+            _CCCLOpKindEnum.STATELESS,
+            self.name,
+            ctypes.c_char_p(self.ltoir),
+            len(self.ltoir),
+            1,
+            1,
+            None,
+        )
+
+
+def _device_array_to_cccl_iter(array):
+    dtype = array.dtype
+    info = _numpy_type_to_info(dtype)
+    # Note: this is slightly slower, but supports all ndarray-like objects as long as they support CAI
+    # TODO: switch to use gpumemoryview once it's ready
+    return _CCCLIterator(
+        1,
+        1,
+        _CCCLIteratorKindEnum.POINTER,
+        _CCCLOp(),
+        _CCCLOp(),
+        info,
+        array.__cuda_array_interface__["data"][0],
+    )
+
+
+def _iterator_to_cccl_iter(it):
+    context = cuda.descriptor.cuda_target.target_context
+    numba_type = it.numba_type
+    size = context.get_value_type(numba_type).get_abi_size(context.target_data)
+    alignment = context.get_value_type(numba_type).get_abi_alignment(
+        context.target_data
+    )
+    (advance_abi_name, advance_ltoir), (deref_abi_name, deref_ltoir) = it.ltoirs.items()
```

Comment:
`it.ltoirs` is a dict, relying on insertion order is to know what is advance and what is deref is a little dicy.

See below.

Reply:
This was definitely true before Python 3.6 but is explicitly legal since Python 3.7. Iterating over a dict view (such as `.values()`) now [returns in insertion order](https://docs.python.org/3/library/stdtypes.html#dictionary-view-objects).

Reply:
I know, but I really wouldn't rely on that. It's too brittle / unobvious; and really buys nothing. Even a tuple of tuples isn't exactly great, but not quite as unobvious.

Also, AFAIK there is no assurance that future versions of Python will guarantee insertion order.